F/golang

bndw opened the pull request:
Replaces the original Python script with Go.

### Why?
* It's more secure. Previously we had to use `subprocess` to invoke GPG for encryption, and that required the master password to be passed in the call.
* It's easier to distribute

### Breaking changes
This implementation changes the data format, therefore encrypted safe files created by the old Python implementation are not compatible with the Go implementation.

I'd like to resolve this with a standalone binary that can update the serialization format of a Python generated safe file. This shouldn't be much more that renaming `data` to `accounts` in the serialized JSON.

leonklingele commented on .gitignore:
```diff
@@ -0,0 +1,2 @@
+.DS_Store
```
This should go into your global `.gitconfig` ;)

leonklingele commented on build:
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/env bash
```
Why not use a `Makefile` instead of a file to `build`, `install`, `clean`, `test` respectively?

bndw replied:
Great question; I chose to not use a Makefile because, as I understand it, it's not idiomatic to use a Makefile in Go projects. I've seen this trend of using bash scripts in [larger](https://github.com/kelseyhightower/confd) projects and thought I'd give it a try.

fwiw, I'm not loving it and very well may drop back to a Makefile.

leonklingele replied:
It's really simple:

$ cat ./Makefile
```Makefile
PWD := $(shell dirname $(realpath $(lastword $(MAKEFILE_LIST))))

GOPKG = github.com/bndw/pick
GOPATH = "$(CURDIR)/vendor:$(CURDIR)"

BIN_DIR = /usr/local/bin
INSTALL = install

all: build

goget:
	GOPATH=$(GOPATH) go get github.com/rogpeppe/godeps
	GOPATH=$(GOPATH) $(CURDIR)/vendor/bin/godeps -u dependencies.tsv
	mkdir -p $(shell dirname "$(CURDIR)/vendor/src/$(GOPKG)")
	rm -f $(CURDIR)/vendor/src/$(GOPKG)
	ln -sf $(PWD) $(CURDIR)/vendor/src/$(GOPKG)

build: goget
	GOPATH=$(GOPATH) go build -o bin/pick .

test: goget
	GOPATH=$(GOPATH) go test -v ./...

install:
	@echo "Installing pick to $(BIN_DIR)/pick"
	$(INSTALL) -c bin/pick $(BIN_DIR)/pick

uninstall:
	rm -f $(BIN_DIR)/pick

clean:
	rm -rf vendor/
	rm -rf bin/

.PHONY: all goget build test install uninstall clean
```

$ cat dependencies.tsv
```
github.com/atotto/clipboard	git	bb272b845f1112e10117e3e45ce39f690c0001ad	2016-02-19T03:44:21Z
github.com/spf13/cobra	git	8e91712f174ced10270cf66615e0a9127e7c4de5	2016-01-17T03:09:38Z
github.com/spf13/pflag	git	7f60f83a2c81bc3c3c0d5297f61ddfa68da9d3b7	2015-12-18T13:47:03Z
golang.org/x/crypto	git	3760e016850398b85094c4c99e955b8c3dea5711	2016-01-13T19:11:26Z
```

- Build via `$ make` / `$ make build`
- Test via `$ make test `
- Install via `$ make install``
- Uninstall via `$ make uninstall`
- Clean via `$ make clean`

leonklingele commented on README.md:
````diff
@@ -4,48 +4,39 @@ A minimal password manager for OS X and Linux.
+or
+
+```sh
+$ git clone https://github.com/bndw/pick && cd pick
````
This only works if you have installed the package with `go get` before.

bndw replied:
Good point. It could probably work with something like
```
git clone https..../pick $GOPATH/src/.../pick
```

bndw commented:
@leonklingele Thanks for reviewing this so far. I'll leave it open for another day or so, should you find anything show stopping.

leonklingele commented on README.md:
````diff
@@ -4,48 +4,30 @@ A minimal password manager for OS X and Linux.
-* **Linux**: `sudo apt-get install gnupg`
+## Install
+```sh
+$ go get github.com/bndw/pick
````
`go get -u` so the package gets updated if you have an outdated version.

leonklingele commented on add.go:
```diff
@@ -0,0 +1,104 @@
+}
+
+func overwrite(name string) bool {
+	prompt := fmt.Sprintf("%s already exists, overwrite (y/n) ?\n", name)
```
Uppercase `N`: `y/N` as `No` is the default.

bndw replied:
I think we should invert that and make `Y` the default. 99% of the time, in my usage, I'm generating passwords.

leonklingele commented on add.go:
```diff
@@ -0,0 +1,104 @@
+	}
+
+	if password == "" {
+		if utils.GetAnswer("Generate password (y/n)?") {
```
Uppercase `N`: `y/N` as `No` is the default.

bndw replied:
I think we should invert that and make `Y` the default. 99% of the time, in my usage, I'm generating passwords.

leonklingele replied:
Sure. Just add another parameter to `GetAnswer` to specify the default value.

leonklingele commented on add.go:
```diff
@@ -0,0 +1,104 @@
+		return "", "", "", 1
+	}
+
+	switch len(args) {
```
Use `fallthrough`:

```go
switch len(args) {
case 3:
	password = args[2]
	fallthrough
case 2:
	username = args[1]
	fallthrough
case 1:
	name = args[0]
}
```

bndw replied:
💯

leonklingele commented on backends/disk.go:
```diff
@@ -0,0 +1,60 @@
+	path string
+}
+
+func defaultSafePath() *string {
```
Why return a string pointer here?

leonklingele commented on backends/disk.go:
```diff
@@ -0,0 +1,60 @@
+const (
+	safeDirMode  = 0700
+	safeMode     = 0600
+	safeFilename = "pick.safe"
```
Add `safeDirName = ".pick"`

leonklingele commented on list.go:
```diff
@@ -0,0 +1,34 @@
+package main
```
Why not put this in a `commands` package?

bndw replied:
Is `commands` a convention for a single executable?

leonklingele replied:
No, I just don't like cluttered project root folders :)

leonklingele commented on safe/load.go:
```diff
@@ -0,0 +1,44 @@
+	"github.com/bndw/pick/utils"
+)
+
+func Load(password []byte, backend backends.Backend) (*Safe, error) {
```
`backends.Backend` -> `Backend`?

bndw replied:
Are you recommending to move the `Backend` interface into the `safe` package?

leonklingele replied:
Forget about this one, sorry. 😿

leonklingele commented on safe/safe.go:
```diff
@@ -0,0 +1,12 @@
+	Accounts map[string]Account `json:"accounts"`
+}
+
+var _password []byte
```
This should not be a global.

bndw replied:
Do you have a recommendation on somewhere else to stick it? It's used again if `safe.Save` is called and I'd prefer not to prompt for it again. A private attribute on `Safe`?

leonklingele replied:
> A private attribute on Safe?

Much better

leonklingele commented on safe/load.go:
```diff
@@ -0,0 +1,44 @@
+		backend = backends.NewDiskBackend(nil)
+	}
+
+	_password = password
```
.. this should not be a global variable

leonklingele commented on util.go:
```diff
@@ -0,0 +1,22 @@
+		return nil, err
+	}
+
+	return safe.Load(password, nil)
```
Return a `nil` error

bndw replied:
`safe.Load` returns `(*safe.Safe, error)`

leonklingele replied:
I'm sorry.

bndw commented:
@leonklingele Currently afk, but I'm happy to accept PR's if you want to contribute any of these changes. Otherwise I'll gladly go through them when I return.

leonklingele commented on utils/crypto.go:
```diff
@@ -0,0 +1,107 @@
+		return password, nil
+	}
+
+	md, err := openpgp.ReadMessage(armorBlock.Body, nil, prompt, nil)
```
Provide a `config` to use AES256 instead of AES128.

bndw replied:
Would you mind opening an issue for this one? If we expose control of encryption, I'd like to consider the design of flags and/or `config` files, ciphers, etc.

leonklingele commented:
Sure. Will open PRs once this one here is merged. :)

leonklingele commented on add.go:
```diff
@@ -0,0 +1,104 @@
+		return handleError(err)
+	}
+
+	fmt.Println("Credential added")
```
This should also output the new password.

bndw replied:
@leonklingele To clarify, are you recommending the `add` command prints the password of the newly added credential to stdout?

leonklingele replied:
Ya, I think it's much better, so you don't have to `add` then `cat`.

leonklingele commented on add.go:
```diff
@@ -0,0 +1,104 @@
+
+func parseAddArgs(args []string) (name, username, password string, errCode int) {
+	if len(args) > 3 {
+		fmt.Println("Usage: add [name] [username] [password]")
```
You should remove the password argument. You probably don't want the password stored in plaintext in your .bash_history ;)

bndw replied:
I prefer to leave the option there, as it may come in handy if it's, for example, called from some other script.

leonklingele commented on remove.go:
```diff
@@ -0,0 +1,14 @@
+	if err := safe.Remove(args[0]); err != nil {
+		return handleError(err)
+	}
+
```
`fmt.Println("Account was removed")`

bndw replied:
sgtm, but it's maintain a consistent grammar- I've been using `credential`.

bndw commented:
Merging so we can carry on.

@leonklingele I believe I've addressed all of your comments here- If I've missed anything please feel free to open an issue.
